e3sm_diags/derivations/formulas_cosp.py

```diff
@@ -0,0 +1,435 @@
+    var_sum = var_sub.sum(dim=[prs.name, tau.name])
+
+    # 6. Set the variable's long name based on the original variable's name and
+    # prs ranges.
+    var_key = str(var.name)
+    simulation = _get_simulation_str(var_key)
+    prs_cloud_level = _get_prs_cloud_level_str(var_key, prs_range, prs_adj_range)
+
+    if simulation is not None and prs_cloud_level is not None:
+        var_sum.attrs["long_name"] = f"{simulation}: {prs_cloud_level} with {tau_lim}"
+
+    # 7. Convert units to %.
+    final_var = convert_units(var_sum, "%")
+
+    return final_var
+
+
+def _get_cloud_axis(var: xr.DataArray, axis: CloudAxis) -> xr.DataArray:
+    da_axis = None
+
+    keys = CLOUD_HIST_MAP[axis]["keys"]
+    for key in keys:
+        if key in var.dims:
+            da_axis = var[key]
+
+            break
+
+    if da_axis is None:
+        raise KeyError(
+            f"The {axis!r} axis is not in the {var.name!r} to "
+            "standardize the cosp histogram."
+        )
+
+    return da_axis
+
+
+def _subset_cloud_var(
+    var: xr.DataArray, prs: xr.DataArray, tau: xr.DataArray
+) -> xr.DataArray:
+    prs_min = CLOUD_HIST_MAP["prs"]["min_mask"]
+    prs_max = prs[-1].item()
+
+    tau_min = CLOUD_HIST_MAP["tau"]["min_mask"]
+    tau_max = tau[-1].item()
+
+    # MISR model and MISR obs
+    if var.name in ["CLD_MISR", "CLMISR"]:
+        # COSP v2 cosp_htmisr in units m instead of km as in v1 and COSP v2
+        # cosp_htmisr[0] equals to 0 instead of -99 as in v1 so the cloud
+        # varable needs to be masked manually by slicing out the first index
+        # on the cosp_htmisr axis.
+        if var.name == "CLD_MISR" and prs.max().item() > 1000:
+            var = var.isel({prs.name: slice(1, None)})
+            prs_max = 1000.0 * prs_max
+
+        cond = (prs >= prs_min) & (prs <= prs_max) & (tau >= tau_min) & (tau <= tau_max)
+    # ISCCP model, # ISCCP obs, and MODIS
+    elif var.name in ["FISCCP1_COSP", "CLISCCP", "CLMODIS"]:
+        cond = (tau >= tau_min) & (tau <= tau_max)
+
+    var_sub = var.where(cond, drop=True)
+
+    return var_sub
+
+
+def _add_missing_cloud_bnds(
+    ds: xr.Dataset, axis_var: xr.DataArray, axis: CloudAxis
+) -> xr.Dataset:
+    # FIXME: ValueError: conflicting sizes for dimension 'misr_cth': length 7
+    # on the data but length 16 on coordinate 'misr_cth'
+    # https://github.com/E3SM-Project/e3sm_diags/issues/761
+    # TODO: Might need to dynamically generate bounds if they don't exist.
+
+    bnds_key = axis_var.attrs.get("bounds")
+    if bnds_key is None or bnds_key not in ds.data_vars.keys():
+        # bnds_data = CLOUD_HIST_MAP[axis]["default_bnds"]
+        # default_bnds = xr.DataArray(
+        #     name=f"{axis_var.name}_bnds",
+        #     data=bnds_data,
+        #     dims=list(axis_var.dims) + ["bnds"],
+        #     coords=axis_var.coords,
+        # )
+        # ds[default_bnds.name] = default_bnds
+        pass
+
+    return ds
```

Comment:
Related to #761

Reply:
I have checked that both the model data and reference data all have bounds for prs and tau. And I don't have a clear idea how to add bounds if they are missing in this case...

Reply:
I will remove this function based on my comment [here](https://github.com/E3SM-Project/e3sm_diags/issues/761#issuecomment-1917557274).